FEATURE: New Review Queue

eviltrout opened the pull request:
This is a large refactor that merges all types of reviewable items in core Discourse (Users, Queued Posts and Flags) into one consolidated review queue.

## New Features

* One convenient place to review all content
* Reviewable items are scored and the highest scored items are shown first. Staff members and higher trust level users score items higher than new users.
* You can filter the review queue by status, score, reviewable type, category
* The API for reviewables returns the list of actions that can be performed on an item, which means the front end application can be simpler and not duplicate all the logic.
* Data structures are built to enable non-staff level moderation of items in the near future

## Software Design Notes

The `Reviewable` model uses single table inheritance (STI), and each type of reviewable must declare a subclass, serializer and ember component. Three types currently exist: `ReviewableQueuedPost`, `ReviewableUser` and `ReviewableFlaggedPost`. Developers are encouraged to look at how other reviewable items are implemented and build their own.

Post Actions (such as flags) must now be created using the `PostActionCreator` service object. It has a much friendlier API now and returns a result object which will return the associated reviewable.

## Backwards Compatibility

REST API is maintained with the old endpoints, however deprecation warnings are logged. Third party code that is pointed at the old APIs should be updated to the new review queue API. All of the Ember.js/front end code for the old user approval / flagging interface / queued posts is removed.

In a future release we will remove the old APIs and backwards compatibility.

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-bundled-action.js.es6:
```diff
@@ -0,0 +1,22 @@
+export default Ember.Component.extend({
+  tagName: "",
+
+  @computed("bundle.actions.length")
```
Can be rewrite with:

```
multiple: Ember.computed.gt("bundle.actions.length", 1)
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+      return _components[type];
+    }
+
+    let dasherized = Ember.String.dasherize(type);
```
```suggestion
    const dasherized = Ember.String.dasherize(type);
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+    let reviewable = this.get("reviewable");
+
+    let performAction = () => {
+      let version = reviewable.get("version");
```
```suggestion
      const version = reviewable.get("version");
```

jjaffeux replied:
```suggestion
      const version = reviewable.get("version");
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+          );
+        })
+        .catch(popupAjaxError)
+        .finally(() => {
```
```
.finally(() => this.set("updating", false))
```

jjaffeux replied:
```
.finally(() => this.set("updating", false));
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+      return _components[type];
+    }
+
+    let dasherized = Ember.String.dasherize(type);
```
```suggestion
    const dasherized = Ember.String.dasherize(type);
```

eviltrout replied:
I don't feel as strongly about making everything `const` as I used to. I find it doesn't catch many bugs, and nothing seems faster as a result (which is initially what I'd hoped.)

jjaffeux replied:
No it mostly doesn't do anything, I agree. In the end all is transpiled into `var`. The only thing it does is hitting the developer about the fact that this is not going to be mutated later, which I do appreciate, but not strong feelings about this.

eviltrout replied:
My current feeling on it is it's up to the developer writing the code. Using `let` or `const` fixes the weird Javascript scope issues, and they should choose what they think is more appropriate. I'll review/approve either form personally.

jjaffeux replied:
works for me :+1:

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    let dasherized = Ember.String.dasherize(type);
+    let template = Ember.TEMPLATES[`components/${dasherized}`];
```
```suggestion
    const template = Ember.TEMPLATES[`components/${dasherized}`];
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+  },
+
+  _performConfirmed(action) {
+    let reviewable = this.get("reviewable");
```
```suggestion
    const reviewable = this.get("reviewable");
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+  _performConfirmed(action) {
+    let reviewable = this.get("reviewable");
+
+    let performAction = () => {
```
```suggestion
    const performAction = () => {
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+    };
+
+    if (action.client_action) {
+      let actionMethod = this[`client${action.client_action.classify()}`];
```
```suggestion
      const actionMethod = this[`client${action.client_action.classify()}`];
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+  },
+
+  _penalize(adminToolMethod, reviewable, performAction) {
+    let adminTools = this.get("adminTools");
```
```suggestion
    const adminTools = this.get("adminTools");
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+  _penalize(adminToolMethod, reviewable, performAction) {
+    let adminTools = this.get("adminTools");
+    if (adminTools) {
+      let createdBy = reviewable.get("target_created_by");
```
```suggestion
      const createdBy = reviewable.get("target_created_by");
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+    let adminTools = this.get("adminTools");
+    if (adminTools) {
+      let createdBy = reviewable.get("target_created_by");
+      let postId = reviewable.get("post_id");
```
```suggestion
      const postId = reviewable.get("post_id");
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+    if (adminTools) {
+      let createdBy = reviewable.get("target_created_by");
+      let postId = reviewable.get("post_id");
+      let postEdit = reviewable.get("raw");
```
```suggestion
      const postEdit = reviewable.get("raw");
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+      let createdBy = reviewable.get("target_created_by");
+      let postId = reviewable.get("post_id");
+      let postEdit = reviewable.get("raw");
+      return this.get("adminTools")[adminToolMethod](createdBy, {
```
```suggestion
      return adminTools[adminToolMethod](createdBy, {
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+        .update(updates)
+        .then(() => this.set("editing", false))
+        .catch(popupAjaxError)
+        .finally(() => {
```
```
.finally(() => this.set("updating", false));
```

jjaffeux commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,162 @@
+        return;
+      }
+
+      let msg = action.get("confirm_message");
```
```suggestion
      const msg = action.get("confirm_message");
```

jjaffeux commented on app/assets/javascripts/discourse/controllers/review-index.js.es6:
```diff
@@ -0,0 +1,58 @@
+        return;
+      }
+
+      let newList = this.get("reviewables").reject(reviewable => {
```
```suggestion
      const newList = this.get("reviewables").reject(reviewable => {
```

jjaffeux commented on app/assets/javascripts/discourse/controllers/review-index.js.es6:
```diff
@@ -0,0 +1,58 @@
+    },
+
+    refresh() {
+      this.set("type", this.get("filterType"));
```
```
this.setProperties({
  type: this.get("filterType"),
  min_score: this.get("filterScore"),
  status: this.get("filterStatus"),
  category_id: this.get("filterCategoryId")
})
```

jjaffeux commented on app/assets/javascripts/discourse/models/reviewable.js.es6:
```diff
@@ -0,0 +1,49 @@
+      return Ember.RSVP.resolve();
+    }
+
+    let adapter = this.store.adapterFor("reviewable");
```
```suggestion
    const adapter = this.store.adapterFor("reviewable");
```

jjaffeux commented on app/assets/javascripts/discourse/routes/review-index.js.es6:
```diff
@@ -0,0 +1,26 @@
+  },
+
+  setupController(controller, model) {
+    let meta = model.resultSetMeta;
```
```suggestion
    const meta = model.resultSetMeta;
```

jjaffeux commented on app/assets/javascripts/discourse/routes/review-topics.js.es6:
```diff
@@ -0,0 +1,11 @@
+  },
+
+  setupController(controller, model) {
+    controller.setProperties({
```
controller.set("reviewableTopics", model);

jjaffeux commented on app/assets/javascripts/select-kit/components/select-kit.js.es6:
```diff
@@ -461,6 +463,7 @@ export default Ember.Component.extend(
     clearSelection() {
       this.deselect(this.get("selection"));
       this.focusFilterOrHeader();
+      this.didClearSelection();
```
❤️

jjaffeux commented on app/assets/stylesheets/common/base/reviewables.scss:
```diff
@@ -0,0 +1,221 @@
+    align-items: center;
+    margin-right: 1em;
+
+    label {
```
as much as possible I prefer when we avoid styling on html elements, it always causes pain later

jjaffeux commented on app/assets/stylesheets/common/base/reviewables.scss:
```diff
@@ -0,0 +1,221 @@
+      margin-left: 0.5em;
+    }
+  }
+  th {
```
same remark on html styles

eviltrout replied:
In this case I think it's more defensible, because a table has a lot of elements th, td that would all need a new class attribute. Instead I made it more explicit by introducing `> th > tr` so that the styles only apply to the format of this table. If there's a nested table they should not apply.

jjaffeux replied:
makes sense :+1:

jjaffeux commented on app/assets/stylesheets/common/base/reviewables.scss:
```diff
@@ -0,0 +1,221 @@
+    max-height: 300px;
+    overflow-y: auto;
+
+    p,
```
html styling

eviltrout replied:
I have no other control over this formatting since it's the `cooked` of a post.

jjaffeux commented on app/assets/stylesheets/common/base/reviewables.scss:
```diff
@@ -0,0 +1,221 @@
+      margin: 0;
+    }
+
+    input[type="text"] {
```
html styling

jjaffeux commented on app/controllers/admin/flags_controller.rb:
```diff
@@ -19,47 +18,39 @@ def index
-        users: serialize_data(users, FlaggedUserSerializer)
+    }
+
+    if (total_rows || 0) > (offset + per_page)
```
Maybe we can store `(offset + per_page)` as we use it later

jjaffeux commented on app/models/report.rb:
```diff
@@ -541,10 +541,19 @@ def self.report_flags(report)
-    countable = PostAction.where(post_action_type_id: PostActionType.flag_types_without_custom.values)
-    countable = countable.joins(post: :topic).merge(Topic.in_category_and_subcategories(report.category_id)) if report.category_id
-    add_counts report, countable, 'post_actions.created_at'
+    basic_report_about(
```
❤️

romanrizzi commented on app/models/reviewable_flagged_post.rb:
```diff
@@ -0,0 +1,284 @@
+    @post ||= (target || Post.with_deleted.find_by(id: target_id))
+  end
+
+  def build_actions(actions, guardian, args)
```
Is there any chance we could move action creation to: `lib/reviewable/actions.rb`.It's not directly related to the actions :thinking:

eviltrout replied:
Can you elaborate? The actions are built dynamically per reviewable type, so these actions are `ReviewableFlaggedPost` specific. You'd have to have `reviewable/actions.rb` test the type of object in some kind of case statement I think.

romanrizzi replied:
This is a pretty complex class that not only perform the review actions but also decides which actions will be visible in the UI. I was asking if we could move some of that complexity away by moving that responsibility to `actions.rb` (if it makes sense)

eviltrout replied:
The actions are used for display in the UI, but it's also used for access checks. For example, if a reviewable is in the `approved` state, it won't return an `approve` action, and if an API request comes in that attempts to review it, it will fail because that action does not exist.

The problem with moving it to actions is then the actions class needs to know about all types of reviewables and how they should respond to things so it could build the appropriate actions.

romanrizzi replied:
That's why I bring this up. This class is taken care of:
  - UI elements
  - Authorization
  - Persistence

and is also an entity of our business domain.

My main intention is just to raise awareness that tight coupling leads to systems that are harder to extend and evolve :stuck_out_tongue_winking_eye:

eviltrout replied:
I think in this case I'm still seeing how you would improve things. If you like, you could create a follow up PR showing how you'd split things up and how it would improve the code.

eviltrout commented on app/controllers/admin/flags_controller.rb:
```diff
@@ -71,65 +62,70 @@ def agree
+        type: 'defer',
+        post: post,
+        user: current_user
+      )
```
That's code I didn't change from before. I do agree it could be put into a method but in this case the entire controller is deprecated and going to be deleted anyway, so probably not worth it.

eviltrout commented on app/controllers/reviewables_controller.rb:
```diff
@@ -0,0 +1,180 @@
+        status: status
+      }
+    }
+    json.merge!(hash)
```
Hmmm I think the `hash` is leftover from some old code I ended up throwing away. I will do a commit here to remove it altogether, thanks.

eviltrout replied:
Oh wait it's not empty. It's populated by the serializer. It's how I set up the side loading of data. It has to stay this way.

romanrizzi replied:
Oh, I see. The hash gets populated as a side effect of the serializer call and that makes it pretty confusing

eviltrout replied:
Yes unfortunately that's the AMS API for handling this. I had to use this hacky way because of STI and I'm not sure there's a better way of doing it.

eviltrout commented on app/controllers/reviewables_controller.rb:
```diff
@@ -0,0 +1,180 @@
+      offset: offset,
+      type: params[:type],
+      min_score: min_score
+    ).to_a
```
I considered that, but also figured you then have to deal with the overhead of a merge. It might be worth it? It was a "borderline" thing for me.

romanrizzi replied:
IMO it depends if both calls are highly connected and the common options are unlikely to be changed only in one place. It's better to avoid duplication when possible

eviltrout replied:
I made this change and it does seem a lot easier to follow, at the minor cost of two extra `merge` calls which should be fairly fast.

eviltrout commented on app/controllers/reviewables_controller.rb:
```diff
@@ -0,0 +1,180 @@
+    category_id = params[:category_id].to_i if params[:category_id]
+
+    topic_id = nil
+    topic_id = params[:topic_id].to_i if params[:topic_id]
```
Thanks, I didn't accept your suggestion here because it can eliminate the `topic_id = nil` line too so I made my own commit.

eviltrout commented on app/controllers/reviewables_controller.rb:
```diff
@@ -0,0 +1,180 @@
+
+  def update
+    reviewable = Reviewable.viewable_by(current_user).where(id: params[:reviewable_id]).first
+    raise Discourse::NotFound.new if reviewable.blank?
```
Good idea, I did so in 8affec2cf. I called it `find_reviewable` to avoid accidentally shadowing it since `reviewable` is a common name.

eviltrout commented on app/models/reviewable.rb:
```diff
@@ -0,0 +1,394 @@
+
+  # subclasses implement "build_editable_fields" to list stuff that can be edited
+  def build_editable_fields(actions, guardian, args)
+  end
```
It doesn't *have* to be overridden though. It's valid to leave it empty when your item is not editable, such as users / flags right now. Instead I've changed the comment to be `subclasses can implement` to make it more clear.

eviltrout commented on lib/flag_query.rb:
```diff
@@ -12,34 +12,26 @@ def self.register_plugin_post_custom_field(field, plugin)
-    if opts[:filter] != "old"
-      post_ids_relation = PostAction.apply_minimum_visibility(post_ids_relation)
-    end
+    total_rows = reviewables.count
```
If you did that you'd get back the count of the rows by `limit / offset` and not the total rows, which is used for pagination.

Also I should mention this is the deprecated API so performance isn't the biggest concern.

romanrizzi replied:
You're right. I Missed that part

eviltrout commented on app/models/reviewable.rb:
```diff
@@ -0,0 +1,394 @@
+require_dependency 'reviewable/perform_result'
+require_dependency 'reviewable_serializer'
+
+class Reviewable < ActiveRecord::Base
```
Can you be more specific? You think all the class methods should be up top and instance methods below for example?

romanrizzi replied:
Exactly. I tend to organize my code that way. what do you think?

eviltrout replied:
It's never been a rule of ours before. I wonder if we should add this to rubocop because it seems like a style change (lots of our code does this.)

romanrizzi replied:
It's definitely something to discuss. This will help us get cleaner interfaces. It's ok to leave this way for now if you want to.

eviltrout replied:
I would suggest a follow up post on our dev or meta forums about this. You can suggest the change and if others are onboard enable the rubocop rule and convert the old code to work.

eviltrout commented on app/models/topic_converter.rb:
```diff
@@ -52,13 +39,24 @@ def convert_to_private_message
+      action_type: [UserAction::GOT_PRIVATE_MESSAGE, UserAction::NEW_PRIVATE_MESSAGE]).find_each do |ua|
+        UserAction.remove_action!(ua.attributes.symbolize_keys.slice(:action_type, :user_id, :acting_user_id, :target_topic_id, :target_post_id))
+      end
+    @topic.posts.each { |post| UserActionManager.post_created(post) }
```
This is a good point, I will make a note about this and will follow up. Topic Conversion was already pretty slow, but on a large topic this could be a time bomb.

eviltrout commented on lib/flag_query.rb:
```diff
@@ -71,68 +62,56 @@ def self.flagged_posts_report(current_user, opts = nil)
+      post = post_lookup[r.target_id]
+      post.post_action_ids ||= []
+
+      r.reviewable_scores.order('created_at desc').each do |rs|
```
`reviewable_scores`  should be preloaded by the `viewable_by` class method already?

romanrizzi replied:
Great! Closing this one

eviltrout commented:
> Sorry for the number of comments 

No problem! Keep them coming. Every suggestion helps out here.

romanrizzi commented:
Will do!

El El dom, 10 mar. 2019 a las 20:57, Robin Ward <notifications@github.com>
escribió:

> *@eviltrout* commented on this pull request.
> ------------------------------
>
> In app/models/reviewable.rb
> <https://github.com/discourse/discourse/pull/7128#discussion_r264066314>:
>
> > @@ -0,0 +1,394 @@
> +require_dependency 'enum'
> +require_dependency 'reviewable/actions'
> +require_dependency 'reviewable/conversation'
> +require_dependency 'reviewable/editable_fields'
> +require_dependency 'reviewable/perform_result'
> +require_dependency 'reviewable_serializer'
> +
> +class Reviewable < ActiveRecord::Base
>
> I would suggest a follow up post on our dev or meta forums about this. You
> can suggest the change and if others are onboard enable the rubocop rule
> and convert the old code to work.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/discourse/discourse/pull/7128#discussion_r264066314>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEywGMFVZUtdqzmUjOVzNTBCGgTwTF1eks5vVZv6gaJpZM4bj6VQ>
> .
>

romanrizzi commented:
I will try to move the actions to the Action class in a follow up PR after
finishing my work on the akismet plugin

El El dom, 10 mar. 2019 a las 20:56, Robin Ward <notifications@github.com>
escribió:

> *@eviltrout* commented on this pull request.
> ------------------------------
>
> In app/models/reviewable_flagged_post.rb
> <https://github.com/discourse/discourse/pull/7128#discussion_r264066257>:
>
> > @@ -0,0 +1,284 @@
> +require_dependency 'reviewable'
> +
> +class ReviewableFlaggedPost < Reviewable
> +
> +  def post
> +    @post ||= (target || Post.with_deleted.find_by(id: target_id))
> +  end
> +
> +  def build_actions(actions, guardian, args)
>
> I think in this case I'm still seeing how you would improve things. If you
> like, you could create a follow up PR showing how you'd split things up and
> how it would improve the code.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/discourse/discourse/pull/7128#discussion_r264066257>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEywGCq4MbNCQCjCf5Y7p5bk5v0p4dRkks5vVZuwgaJpZM4bj6VQ>
> .
>

eviltrout commented:
Okay I've squashed all the commits, and rebased on top of `master` with all the feedback above. There's still time for more feedback as the target merge day would be Wednesday morning.

pmusaraj commented:
I pulled this locally, and noticed some issues while playing around with it (mostly with queued posts): 

1. When a new user posts a reply (site is configured to queue all posts from new users), and an admin approves the reply, the admin is displayed as the poster of the reply. This happens on replies and new topics. 

2. The count of pending posts for a user doesn't reset after approvals. For example, I submitted several test replies and posts as a new user, and approved some of them by the admin account, but the count in the modal keeps incrementing. 

3. There is a delay before the red badge notification shows up for admins. (This might be a local issue, though I don't see a delay on master locally.)

romanrizzi commented:
Looks like the status it's not changing :thinking: 

```ruby
spam_reporter = Discourse.system_user
post = Fabricate(:post)
reviewable = ReviewableFlaggedPost.needs_review!(target: post, topic: post.topic, created_by: spam_reporter)
reviewable.status # 0
reviewable.agree(spam_reporter, {})
reviewable.reload.status # 0
```

Am I doing something wrong?

eviltrout commented:
Thanks Penar, some really good finds. 

1 + 2 have been fixed, legit bugs.

3. I believe is because I am now updating those counts asynchronously. The issue here is previously we could broadcast to all staff members, but with non-staff members approving I needed to do some more logic. I'm hoping it'll be faster outside of dev mode or otherwise not a problem. Let's keep an eye on it.

tgxworld commented on app/serializers/queued_post_serializer.rb:
```diff
@@ -1,36 +1,74 @@
+      .last&.created_by_id
+  end
+
+  #
```
Stray `#` here :grin:

tgxworld commented on app/serializers/queued_post_serializer.rb:
```diff
@@ -1,36 +1,74 @@
+  def post_options
+    json = object.payload.dup
+    json.delete('raw')
+    json
```
You can use `Hash#except` in the future :)

eviltrout replied:
nice tip, following up with a fix.

tgxworld commented on app/models/post.rb:
```diff
@@ -462,8 +456,49 @@ def active_flags
+    self.hidden = true
+    self.hidden_at = Time.zone.now
+    self.hidden_reason_id = reason
+    save
```
```suggestion
    save!
```

tgxworld commented on app/models/reviewable.rb:
```diff
@@ -0,0 +1,395 @@
+      created_at: created_at || Time.zone.now
+    )
+
+    update(score: self.score + rs.score, latest_score: rs.created_at)
```
```suggestion
    update!(score: self.score + rs.score, latest_score: rs.created_at)
```

tgxworld commented on app/models/reviewable.rb:
```diff
@@ -0,0 +1,395 @@
+    )
+
+    update(score: self.score + rs.score, latest_score: rs.created_at)
+    topic.update(reviewable_score: topic.reviewable_score + rs.score) if topic
```
```suggestion
    topic.update!(reviewable_score: topic.reviewable_score + rs.score) if topic
```

tgxworld commented on app/models/reviewable.rb:
```diff
@@ -0,0 +1,395 @@
+#  created_at              :datetime         not null
+#  updated_at              :datetime         not null
+#
+# Indexes
```
Do we need indexes for `topic_id`, `category_id`, `created_by_id`, `reviewable_by_group_id` and `claimed_by_id`? It seems like some of this columns are used for filtering so an index is going to speed up the query once the reviewables table grows.

eviltrout replied:
@tgxworld Good question! I didn't add them because they are generally applied *after* all the other indexes that are more important. 

Almost every time you view the reviewable queue you are viewing pending (status) ordered by (score DESC). On most forums that would return less than 10 records I'd imagine unless the mods were absent, so the other filters needn't have indexes.

There might be a case where we find all the reviewables for a topic that I've forgotten, but honestly I'd prefer to deploy experimentally and watch the DB stats and add indexes as needed, rather than adding indexes to columns that might not need them.

tgxworld commented on app/models/reviewable_flagged_post.rb:
```diff
@@ -0,0 +1,292 @@
+      action.deferred_at = Time.zone.now
+      action.deferred_by_id = performed_by.id
+      # so callback is called
+      action.save
```
```suggestion
      action.save!
```

tgxworld commented on app/models/reviewable_flagged_post.rb:
```diff
@@ -0,0 +1,292 @@
+      action.disagreed_at = Time.zone.now
+      action.disagreed_by_id = performed_by.id
+      # so callback is called
+      action.save
```
```suggestion
      action.save!
```

tgxworld commented on app/models/reviewable_history.rb:
```diff
@@ -0,0 +1,31 @@
+#  reviewable_id           :integer          not null
+#  reviewable_history_type :integer          not null
+#  status                  :integer          not null
+#  created_by_id           :integer          not null
```
I think we need an index here as well.

tgxworld commented on app/models/reviewable_queued_post.rb:
```diff
@@ -0,0 +1,132 @@
+
+    payload['created_post_id'] = created_post.id
+    payload['created_topic_id'] = created_post.topic_id unless topic_id
+    save
```
```suggestion
    save!
```

tgxworld commented on app/models/reviewable_queued_post.rb:
```diff
@@ -0,0 +1,132 @@
+      delete_as_spammer: true
+    }
+
+    if Rails.env.production? && ENV["Staging"].nil?
```
I'm curious about this conditional here. Is there a specific use case for this? My worry here is that the ENV is very generic and its use case isn't documented.

eviltrout replied:
This isn't new code, it's old Queued Posts code that I migrated over. I looked back and I wrote that code many years ago. It's possible we don't need it but it's probably safer to leave for now.

tgxworld replied:
Ah I’m quite sure we can nuke this code now. I don’t think we have any clusters with the Staging ENV. Plus, I would consider this a private API so we don’t have to worry about deprecating.

eviltrout replied:
Sure thing, nuked in an upcoming commit.

tgxworld commented on app/models/reviewable_score.rb:
```diff
@@ -0,0 +1,86 @@
+#  created_at            :datetime         not null
+#  updated_at            :datetime         not null
+#
+# Indexes
```
I feel like most of the `*_id` here will benefit from an index.

eviltrout replied:
There is a performance penalty on indexing columns where inserts become slower. I would rather show that we are using those columns for a particular thing rather than adding indexes on everything. I am not as certain that we will filter a score based on who reviewed it, for example. You'll almost always want to show a score associated with a reviewable (which has an index.)

tgxworld replied:
I guess I’ve always preferred to add indexes to foreign key columns since they are very commonly used in filters and we tend to forget about the index when adding filters in the future. My view is that the overhead on insertion is there but not expensive though. Most of the problems I’ve seen with slow DB queries are usually due to missing indexes on read queries and almost never on write. However, whether we have the additional indexes here or not is not a show stopper :P

eviltrout replied:
This sounds like a dev topic to me "Should we index all foreign key columns?" :)

For now I've been watching the perf closely (and will continue to do so after merge) and I believe we will catch any missing indexes. If you want to open that dev topic was can make it a bigger discussion which GitHub is not ideal for!

tgxworld commented on app/models/web_hook.rb:
```diff
@@ -55,6 +55,10 @@ def self.enqueue_hooks(type, event, opts = {})
 
   def self.enqueue_object_hooks(type, object, event, serializer = nil)
+    if type == :flag
+      Discourse.deprecate("The flags webhook is deprecated. Please use reviewable instead.")
```
Wouldn't this end up spamming our logs in production since it is still being used? Also, I think it'll be helpful to include the version that we're removing this method in.

eviltrout replied:
I'm not sure how many people are actively using this particular web hook, but we should very quickly migrate them to the alternative. I would personally like to see the logs and take that responsibility on.

SamSaffron replied:
I think deprecate is pretty safe it has callsite based protection so the flooding should not be severe.

tgxworld commented on db/migrate/20190110201340_remove_google_plus_from_share_links_site_setting.rb:
```diff
@@ -8,6 +8,6 @@ def up
 
   def down
-    raise ActiveRecord::IrreversibleMigration
+    # raise ActiveRecord::IrreversibleMigration
```
I think this shouldn't be commented?

eviltrout replied:
Oops, I had to do that temporarily to re-build my database in dev. Fixed in an upcoming commit. Good catch.

tgxworld commented on lib/tasks/user_actions.rake:
```diff
@@ -2,13 +2,38 @@ desc "rebuild the user_actions table"
+      UserActionManager.post_action_destroyed(i)
+    end
+  end
+  Topic.all.each { |i| UserActionManager.log_topic(i) }
```
```suggestion
  Topic.all.find_each { |i| UserActionManager.log_topic(i) }
```

tgxworld commented on lib/tasks/user_actions.rake:
```diff
@@ -2,13 +2,38 @@ desc "rebuild the user_actions table"
-  PostAction.all.each { |i| UserActionCreator.log_post_action(i) }
-  Topic.all.each { |i| UserActionCreator.log_topic(i) }
-  Post.all.each { |i| UserActionCreator.log_post(i) }
+  PostAction.all.each do |i|
```
```suggestion
  PostAction.all.find_each do |i|
```

tgxworld commented:
@eviltrout I left some minor comments after reading through the code. The PR is huge so it was more like a skim rather than a review. My only worry here is that we're missing indexes on a couple of relational columns which are most likely going to be used for filtering purposes.

romanrizzi commented on app/jobs/regular/topic_action_converter.rb:
```diff
@@ -0,0 +1,18 @@
+  # Re-creating all the user actions could be very slow, so let's do it in a job
+  # to avoid a N+1 query on a front facing operation.
+  def execute(args)
+    topic = Topic.find_by(id: args[:topic_id])
```
I was wondering if an includes `:posts` will help us here. We only moved the N+1 issue to sidekiq

eviltrout replied:
By default in Rails that would not change anything, because it would select all the posts in the `.posts` method. The real N+1 concern here was that for every post we are creating a new user history object.

eviltrout commented:
Okay, rebased, force pushed up latest batch of fixes. Thanks for excellent feedback @tgxworld, @pmusaraj, @jjaffeux, @romanrizzi - I still plan on merging it tomorrow morning but there's still time for more tweaks.

ZogStriP commented on app/assets/javascripts/admin/templates/users-list-show.hbs:
```diff
@@ -84,15 +67,10 @@
-                {{/if}}
-              </td>
+            {{#if siteSettings.must_approve_users}}
+              <td>{{i18n-yes-no user.approved}}</td>
```
😍

ZogStriP commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,159 @@
+    let dasherized = Ember.String.dasherize(type);
+    let templatePath = `components/${dasherized}`;
+    let template = Ember.TEMPLATES[`${templatePath}`] || Ember.TEMPLATES[`javascripts/${templatePath}`];
+    if (template) {
```
```suggestion
    _components[type] = template ? dasherized : null;
    return _components[type];
```

ZogStriP commented on app/assets/javascripts/discourse/components/reviewable-item.js.es6:
```diff
@@ -0,0 +1,159 @@
+  _penalize(adminToolMethod, reviewable, performAction) {
+    let adminTools = this.get("adminTools");
+    if (adminTools) {
+      let createdBy = reviewable.get("target_created_by");
```
I wonder if `getProperties` is faster?

eviltrout replied:
I don't think it's faster when you consider it has to be de-structured too since the result is a JS object. Also it's been a while since I reviewed `getProperties` but previously it was a for loop on the properties, which is certainly slower than a few `get`s!

ZogStriP commented on app/assets/javascripts/discourse/helpers/editable-value.js.es6:
```diff
@@ -0,0 +1,17 @@
+export function formatCurrency([reviewable, fieldId]) {
```
Why this name? Are we dealy with currencies?

ZogStriP commented on app/assets/javascripts/discourse/initializers/subscribe-user-notifications.js.es6:
```diff
@@ -21,18 +21,12 @@ export default {
+      bus.subscribe("/reviewable_counts", data => {
+        user.set("reviewable_count", data.reviewable_count);
+        if (data.reviewable_count > 0) {
+          user.set("show_reviewables", 1);
```
What does `1` mean? Is `show_reviewables` used anywhere?

ZogStriP commented on app/assets/javascripts/discourse/templates/components/reviewable-queued-post.hbs:
```diff
@@ -0,0 +1,36 @@
+
```
Empty line 🤷‍♂️

ZogStriP commented on app/assets/javascripts/discourse/templates/review-topics.hbs:
```diff
@@ -0,0 +1,42 @@
+</h3>
+
+{{#if reviewableTopics}}
+	<table class='reviewable-topics'>
```
Identation is weird

ZogStriP commented on app/controllers/post_actions_controller.rb:
```diff
@@ -48,21 +36,25 @@ def create
+    if result.failed?
+      render_json_error(result)
+    else
+      render_post_json(result.post, _add_raw = false)
```
I know it's not your change, but we should probably move the `add_raw` optional parameter to a named parameter?

eviltrout replied:
Sure it's a small thing so I'll add it.

ZogStriP commented on app/controllers/reviewables_controller.rb:
```diff
@@ -0,0 +1,164 @@
+    offset = params[:offset].to_i
+
+    if params[:type].present?
+      unless ['ReviewableUser', 'ReviewableQueuedPost', 'ReviewableFlaggedPost'].include?(params[:type])
```
Should we extract that array into a constant so that it's not re-created every time this method is called?

romanrizzi replied:
discourse-akismet will add a new Class called `ReviewableAkismetPost`. We need a way to discover this dynamically :thinking:

eviltrout replied:
I'll add a helper method for this shortly after we merge.

gschlager commented on app/assets/javascripts/discourse/helpers/format-score.js.es6:
```diff
@@ -0,0 +1,5 @@
+import { registerUnbound } from "discourse-common/lib/helpers";
+
+registerUnbound("format-score", function(score) {
+  return (score || 0).toFixed(1);
```
This number should be localized:

```suggestion
  return I18n.toNumber(score || 0, { precision: 1 });
```
